cumulus/client/consensus/aura/src/collators/slot_based/slot_timer.rs

```diff
@@ -173,6 +191,115 @@ where
+			return Err(())
+		};
+
+		// Compute when the next different slot starts.
+		let next_different_slot_time =
+			last_slot_timestamp.as_millis() + slot_duration.as_millis() as u64;
+		let remaining_millis = next_different_slot_time.saturating_sub(now as u64);
+		let next_aura_slot =
+			Slot::from_timestamp(Timestamp::from(next_different_slot_time as u64), slot_duration);
+
+		Ok((Duration::from_millis(remaining_millis as u64), next_aura_slot))
+	}
+
+	/// Adjust the authoring duration to fit into the slot timing.
+	///
+	/// Returns the adjusted authoring duration and the slot that it corresponds to.
+	pub fn adjust_authoring_duration(
+		&mut self,
+		mut authoring_duration: Duration,
+	) -> Option<Duration> {
+		let Ok((duration, next_slot)) = self.time_until_next_slot() else {
+			tracing::error!(
+				target: LOG_TARGET,
+				"Failed to fetch slot duration from runtime. Using unadjusted authoring duration."
+			);
+			return Some(authoring_duration);
+		};
+
+		let Ok((next_duration_change, next_slot_change)) =
+			self.compute_time_until_next_slot_change()
+		else {
+			tracing::error!(
+				target: LOG_TARGET,
+				"Failed to compute time until next slot change. Using unadjusted authoring duration."
+			);
+			return Some(authoring_duration);
+		};
+
+		// The authoring of blocks must stop 1 second before the slot ends.
+		let deadline = next_duration_change.saturating_sub(BLOCK_PRODUCTION_ADJUSTMENT_MS);
+		tracing::debug!(
+			target: LOG_TARGET,
+			?authoring_duration,
+			?duration,
+			last_reported_slot = ?self.last_reported_slot,
+			?next_slot,
+			?next_duration_change,
+			?next_slot_change,
+			?deadline,
+			"Adjusting authoring duration for slot.",
+		);
+
+		// Ensure no blocks are produced in the last second of the slot,
+		// regardless of authoring duration.
+		if deadline == Duration::ZERO {
+			tracing::debug!(
+				target: LOG_TARGET,
+				?next_duration_change,
+				?next_slot_change,
+				"Not enough time left in the slot to adjust authoring duration. Skipping block production for the slot."
+			);
+
+			return None;
+		}
+
+		// Clamp the authoring duration to fit into the slot deadline.
+		// For most cases, the deadline is farther in the future than the authoring duration.
+		if authoring_duration >= deadline {
+			authoring_duration = deadline;
+
+			// Ensure we are not going below the minimum interval within a reasonable threshold.
+			// For 12 cores, we might have a scenario where the last 3 blocks are skipped:
+			// - Block 10: next slot change in 1.493s:
+			// 	 - After adjusting the deadline: 1.493s - 1s = 0.493s the block could be produced
+			//     without issues.
+			// - Block 11: next slot change in 0.993s - skipped by the deadline
+			// - Block 12: next slot change in 0.493s - skipped by the deadline
+			if authoring_duration <
+				BLOCK_PRODUCTION_MINIMUM_INTERVAL_MS.saturating_sub(BLOCK_PRODUCTION_THRESHOLD_MS)
```

Comment:
I think we can be more generous and increase `BLOCK_PRODUCTION_THRESHOLD_MS` to 100ms or something. I introduced the original `BLOCK_PRODUCTION_THRESHOLD_MS` not because blocks below 500ms are not viable at all, but because I wanted to prevent that someone assigns 36 cores to their chain and we arrives at <200ms blocks. So I chose this arbitrary value that we only produce blocks  every 500ms for now. 

In our case here, producing a 400ms block is also okay.